Add initial pinctrl implementation for RT1062

danieldegrasse opened the pull request:
This PR is based off #41448, with the below differences:
- SOC level nodes are used to define each pinctrl option, and overrides are provided at the board level
- pinctrl properties renamed to align with pinctrl node properties in zephyr (where possible)

As with #41448, the nxp lpuart and i2s drivers are enabled. This PR was tested with the shell example.

hakehuang committed:
dts: pinctrl: add dts binding for nxp mcux rt1xxx

add dts binding dedicated for mcux iomuxc settings

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
dts: binding: add pinctrl node to sai and lpuart

enable pinctrl in i2s and lpuart driver dts

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
soc: kconfig: pinctrl select in rt1xxx series

select pinctrl by default in rt series

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
dts: binding: add pinctrl binding header files

add pinctrl_soc required headers

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
dts: pinctrl: add pinctrl and gpr for rt1xxx

add dtsi settings for rt series
dtsi use gpr to replace pinmux
nxp iomuxc has gpr which has more settings than mux and io settings
current solution is to export gpr separately and access then directly

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
board: pinctrl: board configs in rt1xxx

enable pin control in board level

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
drivers: pinctrl: add mcux_rt pinctrl driver

add pinctrl driver for rt1xxx

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
driver: lpuart: add pinctrl support in mcux lpuart

enable pinctrl in lpuart

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
driver: sai: add pinctrl support in mcux sai

enable i2s pinctrl

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
dts: binding: rename pinmux to gpr

rename pinmux to gpr
different from pinmux and io settings gpr will do more IO
settings.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

danieldegrasse committed:
west.yml: Update to pull in pinctrl header from NXP HAL

NXP hal will define constants for pin mux options in RT pinctrl
implementation. Update hal revision to pull in header file.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+        description: |
+          An array of values defining the configuration register for a given
+          pinmux mode. See fsl_iomuxc.h in the NXP HAL for a list.
+      nxp,software_input_on:
```
why not standard property `input-enable`? Btw, this property is behaving as a boolean, so should be `bool`, and naming should use dashes, not underscores.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+        description: |
+          00 Hyst. Disable Field
+          01 Hyst. Enable Field
+      nxp,pull_up_down_config:
```
please, use: `bias-pull-down`. List of options seems to only list a single pull-down resistor value.

hakehuang replied:
as in RT the two options is set to one set of bits, IMHO, keep the naming stupid and simple is better, rather than to fit into a enforced rules which just make the code hard to read.

gmarull replied:
This topic was largely discussed and was done for a reason: standardize as much as possible (remember Zephyr is multiplatform) and to align with other projects (Linux) that already have gone through this problem. I'm afraid we can't now make exceptions because a vendor _doesn't like it_. Of course, you can always take the time to make an alternative proposal if you think the current solution needs to be improved.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+        description: |
+          00 Pin input follows peripheral
+          01 Pin is forced to input
+      nxp,hysteresis_enable:
```
may `input-schmitt-enable` be suitable? Again, it should be a boolean

hakehuang replied:
as user, when I see below two options
```
      input-schmitt-enable:
        required: false
        type: boolean
        description: enable schmitt-trigger mode

      input-schmitt-disable:
        required: false
        type: boolean
        description: disable schmitt-trigger mode
```
I am quite confused on this, and I prefer to keep this as RT user manual name this fucntion.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+        description: |
+          0 PUE_0_Keeper — Keeper
+          1 PUE_1_Pull — Pull
+      nxp,drive_strength:
```
standard `drive-strength` can be used (you can customize it as ST/GD does for `slew-rate`)

hakehuang replied:
it is already customried, and no other platfrom can reuse this rule. I would rather to have NXP dedicated naming, if it is only apply to NXP platform. but leave @danieldegrasse to decide.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+          01 SPEED_1_medium_100MHz_ — medium(100MHz)
+          10 SPEED_2_medium_100MHz_ — medium(100MHz)
+          11 SPEED_3_max_200MHz_ — max(200MHz)
+      nxp,open_drain:
```
standard `drive-open-drain` can be used

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+        description: |
+          0 ODE_0_Open_Drain_Disabled — Open drain disabled
+          1 SRE_1_Open_Drain_Enabled — Open drain enabled
+      nxp,slew_rate:
```
standard `slew-rate` can be used

hakehuang replied:
standard slew-rate is enalge and disable, but in RT the difinitions are Fast and slow, and the naming is confused if use standard.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+          101 DSE_5_R0_5 — R0/5
+          110 DSE_6_R0_6 — R0/6
+          111 DSE_7_R0_7 — R0/7
+      nxp,speed:
```
ST uses `slew-rate` for that, even though you already seem to have a separate slew rate property.

hakehuang replied:
I guess this can not be aligned as there is dedicatd HW setting differences.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+          - 6
+          - 7
+        description: |
+          000 DSE_0_output_driver_disabled_ — output driver disabled
+          001 DSE_1_R0_150_Ohm___3_3V__260_Ohm_1_8V_ — R0(150 Ohm @ 3.3V, 260 Ohm@1.8V)
+          010 DSE_2_R0_2 — R0/2
+          011 DSE_3_R0_3 — R0/3
+          100 DSE_4_R0_4 — R0/4
+          101 DSE_5_R0_5 — R0/5
+          110 DSE_6_R0_6 — R0/6
+          111 DSE_7_R0_7 — R0/7
```
please be a bit more descriptive in help fields

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,143 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  The node has the 'pinctrl' node label set in MCUX RT SoC's devicetree. These
+  node can be autogenerated using the MCUXpresso config tools combined with
+  the parse-rt-mex.py script in the NXP HAL.
```
Would be good to extend this help, see for example ST/GD/Nordic compatible files.

gmarull commented on dts/bindings/i2s/nxp,mcux-i2s.yaml:
```diff
@@ -5,7 +5,7 @@ description: NXP mcux SAI-I2S controller
 compatible: "nxp,mcux-i2s"
 
-include: [i2s-controller.yaml]
+include: [i2s-controller.yaml, pinctrl-device.yaml]
```
suggestion here: in general you'll need at least `pinctrl-0` and `pinctrl-names`, so making them required is a good idea (will generate a better error message), ST does that for example.

gmarull commented on boards/arm/mimxrt1050_evk/mimxrt1050_evk_qspi_defconfig:
```diff
@@ -13,3 +13,4 @@ CONFIG_SERIAL=y
 CONFIG_GPIO=y
 CONFIG_ARM_MPU=y
 CONFIG_HW_STACK_PROTECTION=y
+CONFIG_PINMUX=y
```
Should be dropped

gmarull commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts:
```diff
@@ -7,6 +7,7 @@
 /dts-v1/;
 
 #include <nxp/nxp_rt1060.dtsi>
+#include "mimxrt1060_evk_pinctrl.dtsi"
```
by convention: `mimxrt1060_evk-pinctrl.dtsi`

gmarull commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts:
```diff
@@ -226,3 +231,7 @@ zephyr_udc0: &usb1 {
 	status = "okay";
 };
+
+&iomuxcgpr {
+	status = "okay";
+};
```
note: if you make `CONFIG_PINCTRL=y` default, I think DT node should also be defaulted to `okay`

gmarull commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk_pinctrl.dtsi:
```diff
@@ -0,0 +1,100 @@
+	lpuart1_default: lpuart1_default {
+		iomuxc_gpio_ad_b0_12_lpuart1_tx {
+			pinmux = <IOMUXC_GPIO_AD_B0_12_LPUART1_TX>;
+			nxp,drive_strength = <0x6>;
+			nxp,hysteresis_enable = <0>;
+			nxp,open_drain = <0>;
+			nxp,pull_keeper_enable = <0x1>;
+			nxp,pull_keeper_select = <0>;
+			nxp,pull_up_down_config = <0>;
+			nxp,slew_rate = <0>;
+			nxp,software_input_on = <0x1>;
+			nxp,speed = <0x2>;
```
suggestion: consider defaulting properties to SoC defaults, so that only required fields have to be defined. In many cases, only a few properties need to be adjusted (at least this is the case for ST/GD/Nordic)

danieldegrasse replied:
Is this possible with pinctrl groups? With nodes like ST uses you can define the majority of the overrides at the SoC level, but with pinctrl groups the definition of the pinctrl node is placed within a board specific group. Default groups could be created for each SOC enabled within Zephyr, but that seems like it would deviate from the two pinctrl methods explained in [Zephyr's documentation](https://docs.zephyrproject.org/latest/guides/pinctrl/index.html).

Of course, there's also the possibility that pinctrl nodes like ST uses are the preferred method. If that's the case, let me know and I can rebuild this PR to use that approach.

gmarull replied:
By defaults I mean defaults in the binding file. If set to SoC defaults, then properties resolve to those values if not defined. For example, assume SoC default for `nxp,speed` is `2`, then just set that property to `2` as default so that you don't need to define that for every group unless it has to be changed to another value.

gmarull replied:
Example: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pinctrl/gd%2Cgd32-pinctrl-af.yaml

danieldegrasse replied:
I can set a "sane" default for in the binding file, but different pins on RT parts have different SOC level configurations at boot, so some pins would get incorrect defaults assigned.

gmarull replied:
You can also have multiple compatibles if parts differ (already happens in some other places, e.g. ST PLL nodes, you can have `st,stm32f4xx-pll-clock`.

gmarull commented on drivers/pinctrl/pinctrl_mcux_rt.c:
```diff
@@ -0,0 +1,46 @@
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+			   uintptr_t reg)
+{
+	uint8_t i;
+
+	/* configure all pins */
+	for (i = 0U; i < pin_cnt; i++) {
```
nit: `uint8_t i` scope can be reduced if put inside `for (uint8_t i...` loop

gmarull commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -1085,6 +1099,8 @@ static const struct i2s_driver_api i2s_mcux_driver_api = {
 #define I2S_MCUX_INIT(i2s_id)						\
 	static void i2s_irq_connect_##i2s_id(const struct device *dev); \
 									\
+	PINCTRL_DT_INST_DEFINE(i2s_id)					\
```
nit: missing `;`

gmarull reviewed (changes requested):
Thanks for this contribution, please check my comments

danieldegrasse commented:
@gmarull I'll respond to your unrelated comments individually, but the nonstandard pinctrl names are a deliberate choice. The reasoning is the following:
- The names chosen here closely follow the hardware's names for configuration registers, which should make it easier for users to understand what these devicetree nodes are selecting at the hardware level without needing to look at the driver code
- These pinctrl property names follow the names used by our MCUXpresso configuration tool, which parses an XML structure with fields like the following:
```
<functional_property_declaration id="software_input_on" name="Software Input On" description="Force the selected mux mode Input path no matter of MUX_MODE functionality">
      <state_declaration id="Disable" name="Disabled" description="Input Path is determined by functionality."/>
      <state_declaration id="Enable" name="Enabled" description="Force input path of this pad."/
</functional_property_declaration>
<functional_property_declaration id="hysteresis_enable" name="Hysteresis enable" description="Hysteresis enable">
      <state_declaration id="Enable" name="Enable" description="Hysteresis enable"/>
      <state_declaration id="Disable" name="Disable" description="Hysteresis Disabled"/>
</functional_property_declaration>
```
- Using these names (or at least preserving a 1:1 mapping between them and Zephyr names) makes auto generation of board `dtsi` files relatively simple

I have a PR open on NXP's HAL [here](https://github.com/zephyrproject-rtos/hal_nxp/pull/133) to enable parsing the same configuration file our MCUXpresso config tool is capable of editing, and generate an output `dtsi` file from the configuration file. Of course users can still hand code their own `dtsi` files, but the configuration tool provides error checking that can speed up the process.

You're correct that many of the properties match 1:1, and I think that the auto generation goal could still be achieved pretty easily by just mapping the property names in the script. I'm more concerned about fields like `bias-pull-up` and `bias-pull-down`, since NXP's IOMUXC does not defined one pull up resistor on each pad, it defines several.

Overall, I don't feel strongly about the property names and am happy to rename them if you feel that is best. My main concern is preserving the ability to auto generate `dtsi` files, primarily to expedite NXP's transition to pinctrl within in tree boards.

gmarull commented:
@danieldegrasse Using standard properties when possible needs to be done, as described in https://docs.zephyrproject.org/latest/guides/pinctrl/index.html#pin-control-drivers. This topic was largerly discussed during pinctrl implementation, and the agreement was to respect this rule to standardize as much as possible. For example, it doesn't make sense to introduce `nxp,open_drain` when you have `bias-open-drain`. In case of pull-up/down, you can always do something like:

```
bias-pull-down;
nxp,pull-down = "100k"; /* only required if default choice (SoC default) is not what you need */
```

danieldegrasse commented:
>  This topic was largely discussed during pinctrl implementation, and the agreement was to respect this rule to standardize as much as possible.

Fair point, I will update this PR to use standard names were possible. Given how important SOC defaults appear to be to using pinctrl, I may end up switching this PR to use pinctrl nodes instead of groups

gmarull commented:
> > This topic was largely discussed during pinctrl implementation, and the agreement was to respect this rule to standardize as much as possible.
> 
> Fair point, I will update this PR to use standard names were possible. Given how important SOC defaults appear to be to using pinctrl, I may end up switching this PR to use pinctrl nodes instead of groups

In such case you may require them to be set explicitely, or, provide nodes with SoC defaults. Be aware of this when pre-defining "defaults" (ST has some issues due to that, like enabling pull-up for I2C by default):

![image](https://user-images.githubusercontent.com/25011557/149600210-d221ade3-b14d-4ebb-adfb-2d6512ed9239.png)

danieldegrasse commented:
> In such case you may require them to be set explicitly, or, provide nodes with SoC defaults. Be aware of this when pre-defining "defaults" (ST has some issues due to that, like enabling pull-up for I2C by default):

Thanks for the heads up- my only goal with defining defaults is to ensure that if a user doesn't override any properties on an pin node, it will not have the register value changed from what it is at startup. I see how defaulting to pull up for i2c without directly documenting it could cause issues.

mmahadevan108 commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -13,10 +13,16 @@
 #include <device.h>
 #include <drivers/uart.h>
 #include <drivers/clock_control.h>
+#ifdef CONFIG_PINCTRL
```
Any reason this is conditionally included.

danieldegrasse replied:
This specific `ifdef` is not required, although it does reduce the total include dependencies when pinctrl is disabled. The remaining `ifdef`s in the driver are required, but do you think this one should be removed?

mmahadevan108 replied:
Since we are transitioning to `PINCTRL` why should we provide an `non-PINCTRL` option?

danieldegrasse replied:
Long term, we should not. Short term, we need to to allow pinctrl to be conditional because boards that do not support it will fail to build using the driver if it requires pinctrl

henrikbrixandersen replied:
As long as you don't have all drivers for the given SoC converted from pinmux to pinctrl, you should keep #ifdefs in the drivers. Otherwise, end-users will end up using a hybrid pinmux/pinctrl solution, where _they_ need to investigate the source code of each enabled driver to see if they need to it up using pinctrl dts or pinmux.c.

danieldegrasse commented:
@gmarull I've rewritten the driver to use standard properties as much as possible, and moved the pinctrl system to use pinctrl nodes defined in NXP's HAL (same approach ST takes). The nodes in NXP's HAL have default properties assigned that correspond to the reset value of the register corresponding to the pin that pinctrl node configures. The only deviation from the standard naming I've applied is overriding `bias-pull-up` to be an enum, to allow the user to configure one of the several pull up resistors NXP supports on each pin.

@hakehuang I think I've captured all the hardware features we support in this implementation- if you look at the documentation in the devicetree yaml binding file it calls out what registers are affected by each pinctrl property. Could you let me know if you think there are additional features of NXP's IP that I'm not capturing here?

mmahadevan108 commented on include/drivers/pinctrl/pinctrl_soc_mcux_rt_common.h:
```diff
@@ -0,0 +1,93 @@
+	Z_PINCTRL_STATES_NAME(node_id)[] = {				\
+		UTIL_LISTIFY(DT_NUM_PINCTRL_STATES(node_id),		\
+			     Z_PINCTRL_STATE_INIT, node_id)		\
+	};
```
`Z_PINCTRL_STATES_DEFINE` is already defined in `./include/drivers/pinctrl.h`

danieldegrasse replied:
Thanks for the heads up, removed in latest push

mmahadevan108 commented on include/dt-bindings/pinctrl/mcux_rt-pinctrl.h:
```diff
@@ -0,0 +1,13 @@
+/*
```
What is the purpose of this file?

danieldegrasse replied:
I believe it was present for future use within the DTS, but it is not needed now. I've removed it.

mmahadevan108 commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -328,15 +329,25 @@ static void enable_mclk_direction(const struct device *dev, bool dir)
+	int err;
+	uint8_t state = 0;
 
+	err = pinctrl_apply_state(dev_cfg->pinctrl, state);
```
Could use `PINCTRL_STATE_DEFAULT` for state.

mmahadevan108 commented on dts/bindings/i2s/nxp,mcux-i2s.yaml:
```diff
@@ -64,6 +64,12 @@ properties:
+
+    pinctrl-names:
+      required: true
+
```
Similar addition can be made to the uart `yaml` file

danieldegrasse replied:
We can't make an addition like this to to the uart `yaml` file until all boards supporting the uart driver also support pinctrl. If we add something like this, it effectively means all in tree boards with the `kinetis-lpuart` peripheral need to support pinctrl

This is because without defined pinctrl nodes for an LPUART node, adding these properties will cause the build to fail. It's the same issue that we have with using the `#ifdef CONFIG_PINCTRL` guards in the actual driver. In the case of the i2s driver, all boards that use it _must_ support pinctrl, so these properties are fine and the `ifdef` guards aren't needed.

hakehuang commented:
> think I've captured all the hardware features we support in this implementation- if you look at the documentation in the devicetree yaml binding file it calls out what registers are affected by each pinctrl property. Could you let me know if you think there are additional features of NXP's IP that I'm not capturing here?

I do not find any missing, your code is comprehensive.

mbolivar-nordic commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,137 @@
+  configuration values to reset values for the SOC register. Individual
+  overrides to pinctrl nodes can be applied at the board level.
+
+  for example, here is an override for the GPIO_AD_B0_12 pad:
```
Nitpick: missing capitalization at beginning of sentence

mbolivar-nordic commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,137 @@
+      slew-rate = <1>;
+  }
+
+  This will select GPIO_AD_B0_12 for use as SCL on LPI2C4, and enable
```
Will it? It naively looks to me like this is configuring the behavior if that GPIO is used as SCL, but it doesn't seem to be actually selecting that pin for this purpose. Is this line correct? Just checking.

mbolivar-nordic reviewed (changes requested):
Thanks for this -- a few minor DTS things from me.

danieldegrasse commented:
@mbolivar-nordic Thanks for the feedback- after some internal discussion we are likely going to use a pin group approach instead of pin node based, as we feel it offers better flexibility with pin states, and aligns better with our linux pinctrl implementation. The PR capturing that implementation is #42238. I'm going to close this PR to avoid any further confusion, sorry about that!

mbolivar-nordic commented:
> The PR capturing that implementation is #42238. I'm going to close this PR to avoid any further confusion, sorry about that!

No worries, thanks for the clarification! I'll wait for that to come out of draft before reviewing
